[Lyndon Lim] iP

lyndonlim27 opened the pull request:
### DukePro

> “Your mind is for having ideas, not holding them.” – David Allen [(source)](https://dansilvestre.com/productivity-quotes)


DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://github.com/lyndonlim27/ip/releases/tag/A-Jar).
2. double-click it and run the eren.jar file
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE!**

Features:

- [x]  Manage Todo tasks
- [x]  Manage Deadline tasks
- [x]  Manage Event tasks
- [x]  Find task using keywords
- [ ]  Managing deadlines (coming soon)
- [ ]  Reminders (coming soon)


If you Java programmer, you can use it to practice Java too. Here's the main method:


```
public static void main(String[] args) throws IOException {
    printDukeIntro();
    boolean hasEnded;
    Storage storage = new Storage();
    TaskList tasks = new TaskList(storage.loadFile());
    Parser parser = new Parser();
    Ui ui = new Ui(parser);

    while(true) {
        hasEnded = ui.receiveInput(tasks, storage);
        if(hasEnded){
            break;
        }
    }
}
```

Commit:
Add Gradle support

lyndonlim27 committed:
Added functionality for Greet, Echo, Exit.

lyndonlim27 committed:
Ensure that Bye is not printed again before exiting

lyndonlim27 committed:
Revert "Ensure that Bye is not printed again before exiting"

This reverts commit efab08b268593b8bdc943d06fc7d76746bd51bc8.

lyndonlim27 committed:
Added the functionality of allowing users to store inputs in a list

lyndonlim27 committed:
Added functionality of marking and unmarking tasks.
Handled some exceptions as well.

lyndonlim27 committed:
Add special task and included inheritance

lyndonlim27 committed:
Add semi-automate testing using I/O redirection technique

lyndonlim27 committed:
Added custom exception and handled more exceptions

lyndonlim27 committed:
Added delete function

lyndonlim27 committed:
Cleaned up and organised codes.
Added more java docs.
Modified a custom exception

lyndonlim27 committed:
Fix indentation

lyndonlim27 committed:
Add save function

lyndonlim27 committed:
Add date and time function

lyndonlim27 committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java

lyndonlim27 committed:
Remove spaces

lyndonlim27 committed:
Merge branch 'branch-Level-7'

lyndonlim27 committed:
Fix merge conflict residues

lyndonlim27 committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

lyndonlim27 committed:
Add gradle to project

lyndonlim27 committed:
Add Ui, Parser, DateTime, Storage and TaskList class

lyndonlim27 committed:
Divide classes into packages

lyndonlim27 committed:
Add JUnit tests

lyndonlim27 committed:
Add javadocs

lyndonlim27 committed:
Tidy code based on coding standards

lyndonlim27 committed:
Add find function for task

lyndonlim27 committed:
Extract out code from main function to a new function

KSunil2001 commented on src/main/java/duke/DateTime.java:
```diff
@@ -0,0 +1,37 @@
+public class DateTime {
+    
+    /**
+     * Return dateTime from user given input
```
Shouldn't it be "Returns" instead of "Return"? 🤔

KSunil2001 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,37 @@
+                + "|  |____  |  |  \\  | |  |____  |    \\  |  |\n"
+                + "|  _____| |  |__/ /  |  _____| |  |\\ \\ |  |\n"
+                + "|  |____  |  | |  \\  |  |____  |  | \\ \\|  |\n"
+                + "|_______| |__| |___\\ |_______| |__|  \\____|\n";
```
I like the creativity in the implementation of your logo 😄

KSunil2001 commented on src/main/java/duke/DateTime.java:
```diff
@@ -0,0 +1,37 @@
+        if (index != -1 && words.length >= index+2) {
+            try {
+                DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd/MM/yyyy HHmm");
+                dateTime = LocalDateTime.parse(words[index+1] + " " + words[index+2], formatter); 
```
I think it would look more neat if there was a space between the "+" operator. 
```suggestion
                dateTime = LocalDateTime.parse(words[index + 1] + " " + words[index + 2], formatter); 
```

KSunil2001 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,92 @@
+            break;
+        default:
+            throw new InvalidCommandException("Invalid command. Please try again");
+        }
```
Good job following the switch case indentation format. It looks very neat 👍

KSunil2001 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,131 @@
+import duke.task.Deadline;
+import duke.task.Event;
+import duke.task.Task;
+import duke.task.Todo;
```
I like how you have grouped the import statements accordingly, but you could take it one step further by spacing the various import statements by their respective groups for better readability.

```suggestion
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;

import java.nio.file.Files;
import java.nio.file.Paths;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

import java.util.ArrayList;

import duke.task.Deadline;
import duke.task.Event;
import duke.task.Task;
import duke.task.Todo;
```

KSunil2001 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,131 @@
+        fileContent = "  TYPE  | COMPLETED | DETAILS | DATE\n";
+        for (Task t : tasks) {
+            dateTime = null;
+            if (t.getType() == 'T') {
+                fileContent += "  Todo  |";
+            }
+            else if (t.getType() == 'D') {
+                fileContent += "Deadline|";
+                if (t instanceof Deadline) {
+                    Deadline task = (Deadline) t;
+                    dateTime = task.getDateTime();
+                }
+            }
+            else {
+                fileContent += "  Event |";
+                if (t instanceof Event) {
+                    Event task = (Event) t;
+                    dateTime = task.getDateTime();
+                }
+            }
```
Should the "else if" and "else" statements start on a new line? 🤔

lyndonlim27 replied:
Thank you! I totally miss that out

KSunil2001 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,142 @@
+     * 
+     * @param num Index of the task to be deleted
+     * @return true if task has been deleted, else false
+     * @throws IndexOutOfBoundsExceptio If num > listOfTask.size()
```
Just pointing out a very minor error. 
```suggestion
     * @throws IndexOutOfBoundsException If num > listOfTask.size()
```

KSunil2001 reviewed (commented):
LGTM, almost. Just a few nits to fix. 
Overall the code looks good, well done😄

winsonheng commented on src/main/java/duke/DateTime.java:
```diff
@@ -0,0 +1,37 @@
+                return dateTime;
+            } catch (DateTimeParseException e) {
+                return null;
+            } catch (IndexOutOfBoundsException e) {
```
Is there a need for this exception if you already checked for index bounds in the if clause?

winsonheng commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,92 @@
+        boolean hasTaskChanged = false;
+        words[0] = words[0].toUpperCase();
+        switch(words[0]) {
+        case "BYE":
```
Would it be better to enumerate the different possible inputs so it's easier to keep track of them and can easily be used elsewhere too?

winsonheng commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,92 @@
+    public boolean runCommand(TaskList tasks, Storage storage, String[] words) throws IndexOutOfBoundsException,
+            NumberFormatException, InvalidCommandException, EmptyDescriptionException {
+
+        boolean hasTaskChanged = false;
```
I like how you keep track of the task changed status so you don't have to write to file with every command!

winsonheng commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,92 @@
+        default:
+            throw new InvalidCommandException("Invalid command. Please try again");
+        }
+        if(words[0] != "LIST" && hasTaskChanged) {
```
Is `words[0] != "LIST"` necessary when the command does not set `hasTaskChanged` to true?

lyndonlim27 replied:
That's right. Thank you!

winsonheng commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,131 @@
+ */
+public class Storage {
+    private static String FILE_PATH = "./data/eren.txt";
+    private static String DIR_PATH = "./data/";
```
Good use of static class fields, but could they be set to final instead?

winsonheng reviewed (commented):
LGTM!
Just some minor nits but good job on your progress so far!

lyndonlim27 committed:
Add GUI

lyndonlim27 committed:
Add assertion for some methods.

Add assertion for the following classes
1) Duke
	- To check mainWindow instance is pass properly to Duke class
	   as it is crucial for displaying the relevant GUI
2) Storage
	- To check number of task load from file is non-negative
3) TaskList
	- To check index return from Parser is greater or equal to -1

lyndonlim27 committed:
Merge pull request #1 from lyndonlim27/branch-A-Assertions

Add assertion for some methods.

lyndonlim27 committed:
Improve code quality and bug fixes

Fix multiple javadoc comment errors. Use normal comment
instead of javadoc comment.

Base on code quality use returns instead of return

Remove unused imports and shorten long condition expressions
by assigning to variables first to prevent confusion

Arrowhead shape code is changed by adding an extra else if
statement and using continue

Fix date conversion as wrong date was shown

Add functionality to automatically exit once user enters "bye"
command

lyndonlim27 committed:
Merge pull request #2 from lyndonlim27/branch-A-CodeQuality

Improve code quality and bug fixes

lyndonlim27 committed:
Add command help and fix GUI

Provide in-App guidance to users by displaying all commands

GUI is resizeable vertically and horizontally

lyndonlim27 committed:
Clean code and add Ui.png

- Abstract out codes into new methods as some methods are too long and have deep nesting
- Change Task to abstract class so that all types of task have the method getDateTime()
- Fix commenting for some classes

lyndonlim27 committed:
Update README.md

lyndonlim27 committed:
Fix bugs

- Text was not displayed when Todo task was added
- Find command was case-sensitive

lyndonlim27 committed:
Revert "Fix bugs"

This reverts commit ea60131b45e119d82bd7ec52cecdf15960dc40cd.

lyndonlim27 committed:
Revert "Revert "Fix bugs""

This reverts commit 2918de9ce9c7c9fa6b77fad2ad180859e4e151b4.

lyndonlim27 committed:
Merge branch 'master' of github.com:lyndonlim27/ip

lyndonlim27 committed:
Update code quality

- Fix code quality after receiving automatic feedback from bot script
